File: 3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0
        end = 0
        n = len(s)
        seen = set()
        ret = 0
        while end < n:
            if ord(s[end]) not in seen:
                seen.add(ord(s[end]))
                end += 1
            else:
                while ord(s[end]) in seen:
                    seen.remove(ord(s[start]))
                    start += 1
            ret = max(ret, end - start)
        return ret
                
        
        
        

File: 3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0;
        int ret = 0;
        Map<Character, Integer> count = new HashMap<>();
        int repeat = 0;
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            count.put(c, count.getOrDefault(c, 0) + 1);
            if(count.get(c) > 1){
                repeat += 1;
            }
            while(repeat > 0){
                char start_c = s.charAt(start);
                start++;
                count.put(start_c, count.get(start_c) - 1);
                if(count.get(start_c) > 0){
                    repeat -= 1;
                }
            }
            ret = Math.max(ret, i - start + 1);
        }
        return ret;
        
    }
}